tests/framework/prototype.py

```diff
@@ -0,0 +1,279 @@
+#   (1) Add (create) CR (2) Update CR (3) Delete CR
+#
+# The member functions integrate together in `trigger()`.
+#   [Step1] exec(): Execute a command to trigger the CREvent. For example, create a CR by a
+#                  `kubectl apply` command.
+#   [Step2] wait(): Wait for the system to converge.
+#   [Step3] checkRuleSets(): When the system converges, check all registered RuleSets.
+class CREvent:
+    def __init__(self, cr, ruleSets: list[RuleSet], timeout, namespace):
+        self.ruleSets = ruleSets
+        self.timeout = timeout
+        self.namespace = namespace
+        self.cr = cr
+    def trigger(self):
+        self.exec()
+        self.wait()
+        self.checkRuleSets()
+    def exec(self):
+        raise NotImplementedError
+    def wait(self):
+        time.sleep(self.timeout)
+    def checkRuleSets(self):
+        for rs in self.ruleSets:
+            rs.checkRuleSet(self.cr, self.namespace)
+
+'''
+My implementations
+'''
+class HeadPodNameRule(Rule):
+    def trigger_condition(self, cr) -> bool:
+        steps = "spec.headGroupSpec".split('.')
+        return (search_path(cr, steps) != None)
+
+    def assertRule(self, cr=None, cr_namespace='default'):
+        expected_val = search_path(cr, "spec.headGroupSpec.template.spec.containers.0.name".split('.'))
+        headpods = client.CoreV1Api().list_namespaced_pod(namespace = cr_namespace, label_selector='rayNodeType=head')
+        assert(headpods.items[0].spec.containers[0].name == expected_val)
+
+class HeadSvcRule(Rule):
+    """The labels of the head pod and the selectors of the head service must match."""
+    def assertRule(self, cr=None, cr_namespace='default'):
+        head_services = client.CoreV1Api().list_namespaced_service(namespace= cr_namespace, label_selector="ray.io/node-type=head")
+        assert(len(head_services.items) == 1)
+        selector_dict = head_services.items[0].spec.selector
+        selector = ','.join(map(lambda key: f"{key}={selector_dict[key]}", selector_dict))
+        headpods = client.CoreV1Api().list_namespaced_pod(namespace =cr_namespace, label_selector=selector)
+        assert(len(headpods.items) == 1)
+
+class EasyJobRule(Rule):
+    """Submit a very simple Ray job to test the basic functionality of the Ray cluster."""
+    def assertRule(self, cr=None, namespace='default'):
+        headpods = client.CoreV1Api().list_namespaced_pod(namespace = namespace, label_selector='rayNodeType=head')
+        headpodName = headpods.items[0].metadata.name
+        rtn = os.system("kubectl exec {} -- python -c \"import ray; ray.init(); print(ray.cluster_resources())\""
+                        .format(headpodName))
+        assert(rtn == 0)
+
+class RayClusterAddCREvent(CREvent):
+    def exec(self):
+        client.CustomObjectsApi().create_namespaced_custom_object(
+            group = 'ray.io',version = 'v1alpha1', namespace = self.namespace, plural = 'rayclusters', body = self.cr,
+            pretty = 'true')
```

Comment:
What is the pretty arg for? Can we remove it?

Reply:
Based on the [doc](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CustomObjectsApi.md#create_namespaced_custom_object), if `pretty` is 'true', then the output is pretty printed. Yes, we can remove it.